File: backend/app/routers/group_code.py
"""
그룹코드 API 라우터
tb_group_code 테이블에 대한 CRUD 엔드포인트를 정의한다.
URL 접두사: /api/v1/group-codes
"""
from fastapi import APIRouter, HTTPException, Query
from typing import Optional

from app.models.common import BaseResponse, PaginatedResponse
from app.models.group_code import GroupCodeCreate, GroupCodeUpdate
from app.services import group_code as group_code_service

router = APIRouter(
    prefix="/api/v1/group-codes",
    tags=["그룹코드"],
)


@router.get("", response_model=PaginatedResponse)
def list_group_codes(
    group_code: Optional[str] = Query(None, description="그룹코드 검색어"),
    group_name: Optional[str] = Query(None, description="그룹명 검색어"),
    page: int = Query(1, ge=1, description="페이지 번호"),
    size: int = Query(20, ge=1, le=100, description="페이지당 항목 수"),
):
    """그룹코드 목록을 페이지네이션하여 조회한다."""
    try:
        rows, total = group_code_service.list_group_codes(
            group_code=group_code,
            group_name=group_name,
            page=page,
            size=size,
        )
        return PaginatedResponse(data=rows, total=total, page=page, size=size)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"그룹코드 목록 조회 실패: {str(e)}")


@router.get("/all", response_model=BaseResponse)
def get_all_group_codes():
    """셀렉트박스용 전체 그룹코드 목록을 조회한다 (삭제되지 않은 항목만)."""
    try:
        rows = group_code_service.get_all_group_codes()
        return BaseResponse(data=rows, message="조회 성공")
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"그룹코드 전체 목록 조회 실패: {str(e)}")


@router.get("/{group_code}", response_model=BaseResponse)
def get_group_code(group_code: str):
    """특정 그룹코드의 상세 정보를 조회한다."""
    try:
        row = group_code_service.get_group_code(group_code)
        if not row:
            raise HTTPException(status_code=404, detail="해당 그룹코드를 찾을 수 없습니다.")
        return BaseResponse(data=row, message="조회 성공")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"그룹코드 조회 실패: {str(e)}")


@router.post("", response_model=BaseResponse, status_code=201)
def create_group_code(body: GroupCodeCreate):
    """새로운 그룹코드를 생성한다."""
    try:
        # 중복 체크
        existing = group_code_service.get_group_code(body.group_code)
        if existing:
            raise HTTPException(status_code=400, detail="이미 존재하는 그룹코드입니다.")

        row = group_code_service.create_group_code(body.model_dump())
        return BaseResponse(data=row, message="등록 성공")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"그룹코드 등록 실패: {str(e)}")


@router.put("/{group_code}", response_model=BaseResponse)
def update_group_code(group_code: str, body: GroupCodeUpdate):
    """기존 그룹코드 정보를 수정한다."""
    try:
        # 존재 여부 확인
        existing = group_code_service.get_group_code(group_code)
        if not existing:
            raise HTTPException(status_code=404, detail="해당 그룹코드를 찾을 수 없습니다.")

        row = group_code_service.update_group_code(
            group_code, body.model_dump(exclude_unset=True)
        )
        return BaseResponse(data=row, message="수정 성공")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"그룹코드 수정 실패: {str(e)}")


@router.delete("/{group_code}", response_model=BaseResponse)
def delete_group_code(group_code: str):
    """그룹코드를 논리 삭제(소프트 딜리트)한다."""
    try:
        existing = group_code_service.get_group_code(group_code)
        if not existing:
            raise HTTPException(status_code=404, detail="해당 그룹코드를 찾을 수 없습니다.")

        row = group_code_service.delete_group_code(group_code)
        return BaseResponse(data=row, message="삭제 성공")
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"그룹코드 삭제 실패: {str(e)}")

File: admin-web/src/api/groupCode.js
/**
 * 그룹코드 API 모듈
 * - 그룹코드 CRUD 관련 API 호출 함수를 정의한다.
 * - 모든 요청은 공통 Axios 인스턴스를 사용한다.
 */
import api from './index'

const BASE_URL = '/api/v1/group-codes'

/**
 * 그룹코드 목록 조회 (페이징/검색 파라미터 포함)
 * @param {Object} params - 검색 조건 (page, size, group_code, group_name 등)
 */
export function getList(params) {
  return api.get(BASE_URL, { params })
}

/**
 * 전체 그룹코드 목록 조회 (셀렉트박스 등에서 사용)
 */
export function getAll() {
  return api.get(`${BASE_URL}/all`)
}

/**
 * 그룹코드 상세 조회
 * @param {string} groupCode - 그룹코드 PK
 */
export function getDetail(groupCode) {
  return api.get(`${BASE_URL}/${groupCode}`)
}

/**
 * 그룹코드 생성
 * @param {Object} data - 생성할 그룹코드 데이터
 */
export function create(data) {
  return api.post(BASE_URL, data)
}

/**
 * 그룹코드 수정
 * @param {string} groupCode - 수정 대상 그룹코드 PK
 * @param {Object} data - 수정할 데이터
 */
export function update(groupCode, data) {
  return api.put(`${BASE_URL}/${groupCode}`, data)
}

/**
 * 그룹코드 삭제
 * @param {string} groupCode - 삭제 대상 그룹코드 PK
 */
export function remove(groupCode) {
  return api.delete(`${BASE_URL}/${groupCode}`)
}
